pkg/lib/indexer/indexer.go

```diff
@@ -573,16 +590,96 @@ type bundleDirPrefix struct {
 	pkgName, bundleVersion string
 }
 
-func getBundlesToExport(dbQuerier pregistry.Query, packages []string) (map[string]bundleDirPrefix, error) {
-	bundleMap := make(map[string]bundleDirPrefix)
+type bundleExportInfo struct {
+	name       string
+	bundlePath string
+	replaces   string
+	skips      map[string]struct{}
+}
 
+func getBundlesToExport(dbQuerier pregistry.Query, packages []string) (map[bundleDirPrefix]bundleExportInfo, error) {
+	bundleMap := make(map[bundleDirPrefix]bundleExportInfo)
+	graphLoader := &sqlite.SQLGraphLoader{
+		Querier: dbQuerier,
+	}
 	for _, packageName := range packages {
-		bundlesForPackage, err := dbQuerier.GetBundlesForPackage(context.TODO(), packageName)
+		bundleGraph, err := graphLoader.Generate(packageName)
 		if err != nil {
 			return nil, err
 		}
-		for k, _ := range bundlesForPackage {
-			bundleMap[k.BundlePath] = bundleDirPrefix{pkgName: packageName, bundleVersion: k.Version}
+		versionMap := make(map[string]string)
+		for _, channel := range bundleGraph.Channels {
+			var replaceDepth map[string]int64
+			for node, replaceNodes := range channel.Nodes {
+				bundleKey := bundleDirPrefix{
+					pkgName:       packageName,
+					bundleVersion: node.Version,
+				}
+				var replaces string
+				var skips map[string]struct{}
+				if bundleInfo, ok := bundleMap[bundleKey]; ok {
+					if bundleInfo.name != node.CsvName || bundleInfo.bundlePath != node.BundlePath {
+						return nil, fmt.Errorf("cannot export bundle version %s for package %s: multiple CSVs found", node.Version, packageName)
+					}
+					replaces = bundleInfo.replaces
+					skips = bundleInfo.skips
+				}
+				if skips == nil {
+					skips = map[string]struct{}{}
+				}
+				if len(replaces) > 0 {
+					skips[replaces] = struct{}{}
+				}
+				for replace := range replaceNodes {
+					if replace.CsvName == replaces {
+						continue
+					}
+					if _, ok := channel.Nodes[replace]; !ok {
+						//synthetic entry
+						skips[replace.CsvName] = struct{}{}
+						continue
+					}
+					if len(replaces) == 0 {
+						replaces = replace.CsvName
+						versionMap[replaces] = replace.Version
+						continue
+					}
+
+					var err error
+					if replaceDepth == nil {
+						replaceDepth, err = dbQuerier.GetBundleReplacesDepth(context.TODO(), packageName, node.CsvName)
+					}
+					if err == nil && replaceDepth[replaces] != replaceDepth[replace.CsvName] {
+						if replaceDepth[replaces] > replaceDepth[replace.CsvName] {
+							replaces = replace.CsvName
+							versionMap[replaces] = replace.Version
+						} else {
+							skips[replace.CsvName] = struct{}{}
+						}
+						continue
+					}
+
+					replacesSemver, replacesErr := semver.Make(versionMap[replaces])
+					newSemver, newErr := semver.Make(replace.Version)
+					if replacesErr == nil && newErr == nil && !replacesSemver.EQ(newSemver) {
```

Comment:
this makes sense but is a bit confusing -- it combines the case where there is an issue with making the semver representation of the version and if the old/new semver representations do not match. it's also a bit repetitive, since if there is an error making a semver type we can almost be sure the types won't match. WDYT about handling the error case explicitly?

Reply:
That makes sense, it would make the error a lot clearer